File: SourceryTests/Stub/JavaScriptTemplates/Equality.js
<% for (type of types.classes) { -%>
extension <%= type.name %>: Equatable {}

<% if (type.annotations.showComment) { _%> // <%= type.name %> has Annotations <%_ } %>

func == (lhs: <%= type.name %>, rhs: <%= type.name %>) -> Bool {
    <% for (variable of type.variables) { -%>
    if lhs.<%= variable.name %> != rhs.<%= variable.name %> { return false }
    <% } -%>
    return true
}
<% } %>

File: SourceryTests/Stub/SwiftTemplates/Equality.swift
import SourceryRuntime

enum EqualityGenerator {
    static func generate(for types: Types) -> String {
        return types.classes.map { $0.generateEquality() }.joined(separator: "\n")
    }
}

extension Class {
    func generateEquality() -> String {
        let propertyComparisons = variables.map { $0.generateEquality() }.joined(separator: "\n    ")

        return """
        extension \(name): Equatable {}
        \(hasAnnotations())
        func == (lhs: \(name), rhs: \(name)) -> Bool {
            \(propertyComparisons)
        
            return true
        }
        
        """
    }

    func hasAnnotations() -> String {
        guard annotations["showComment"] != nil else {
            return ""
        }
        return """

        // \(name) has Annotations
        
        """
    }
}

extension Variable {
    func generateEquality() -> String {
        return "if lhs.\(name) != rhs.\(name) { return false }"
    }
}
